IMDS: Added tagsList field and removed Gallery Application provider

tushar-bhargava-ms opened the pull request:
This PR retroactively introduces the `tagsList` field in the Compute provider part of the spec. This field had accidentally been omitted from the Swagger spec.

This PR also removes the Gallery Application provider part of the spec as that provider is only in private preview, and, consequently, shouldn't be part of our public Swagger spec. (More pressingly, it was also causing the merge validation checks to fail.)

As a slight aside, there are several commits in this PR because of incremental work, but you should squash them on your side before merging for a less scattered commit history!

### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.

tushar-bhargava-ms committed:
Adding missing Swagger spec.

tushar-bhargava-ms committed:
Merge remote-tracking branch 'upstream/master'

tushar-bhargava-ms committed:
Updated README.md with reference to added directory.

tushar-bhargava-ms committed:
Fixing merge conflicts.

tushar-bhargava-ms committed:
Added Swagger spec for IMDS API version 2019-06-04.

tushar-bhargava-ms committed:
Updated 2019-08-01 version with tagsList field.

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-python

tushar-bhargava-ms committed:
Fixing whitespace issue.

azuresdkci commented:
Can one of the admins verify this patch?

Juliehzl commented:
@tushar-bhargava-ms Pls do some changes with examples according to CI error message.

tushar-bhargava-ms committed:
Fixed validation errors.

tushar-bhargava-ms committed:
Fixing more validation errors.

tushar-bhargava-ms committed:
Added schema definition for 204 Gallery app response.

tushar-bhargava-ms committed:
Fixed validation errors in response.

tushar-bhargava-ms committed:
Added empty schema object to 204 responses.

tushar-bhargava-ms commented:
@Juliehzl, I've fixed the validation errors. Sorry for not fixing them earlier: I had contacted the Gallery App team about them and was waiting on their reply.

johanste commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2019-08-01/imds.json:
```diff
@@ -77,6 +81,10 @@
           },
+          "204": {
+            "description": "No application profiles found.",
+            "schema": {}
```
Assuming that you actually don't return anything, the `schema` key should be omitted.

tushar-bhargava-ms replied:
Agreed. However, when I tried to leave out the `schema` field, the merge validation checks failed, complaining about the missing schema. I settled on this as the workaround for the time being, but I'm happy to revert my change if the merge validation checks are updated accordingly.

tushar-bhargava-ms replied:
Nick's suggestion helped me fix this. I've removed the `schema` key.

johanste commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2019-08-01/imds.json:
```diff
@@ -37,6 +37,10 @@
               "$ref": "#/definitions/GalleryApplications"
             }
           },
+          "204": {
```
Why not return a gallery application with an empty applications list? I assume that this is what happened in previous API version. This change introduces friction for clients that want to adopt the new API version.

tushar-bhargava-ms replied:
This is correcting a previously incorrect Swagger. The 204 response was always a part of the API.

I'm unsure. IMDS has several providers that are maintained by other teams. In this case, the Gallery Application team would be best suited to answer that question. @koallison what do you think of returning an empty applications list?

koallison replied:
The Gallery Application API is new, so there aren't any users of Gallery Application using old API versions.

nschonni commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2019-06-04/examples/GetGalleryApplication.json:
```diff
@@ -0,0 +1,15 @@
+            "body": "{application binary package}"
+        },
+        "204": {
+            "body": ""
```
This should be removed instead of adding the `"schema": {}`

tushar-bhargava-ms replied:
I hadn't thought of that! That's much cleaner; I'll make the change.

nschonni replied:
No worries, I had just been cleaning those up on some other Swaggers files since it didn't used to get caught by OAV

tushar-bhargava-ms committed:
Removing schema from 2019-08-01 version and checking validation.

tushar-bhargava-ms committed:
Removed remaining references to empty schema object.

tushar-bhargava-ms commented:
@Juliehzl I've fixed the syntactical errors in the Swagger and made modifications according to the suggestions made by the Azure API board. Could you please review and approve this Swagger?

tushar-bhargava-ms committed:
Rebasing with fixed merge conflict.

tushar-bhargava-ms commented:
@Juliehzl  Checking in on this again :)

tushar-bhargava-ms commented:
@Juliehzl @johanste I'd really appreciate it if you could review the PR and merge it. It's been a little over three weeks since I opened the PR, and our customers are waiting on the newest Swagger specs.

johanste commented:
I am confused by the PR. The title says that you are adding descriptions (for an old API version). However, you are also modifying descriptions for other API versions (e.g. adding the 204 response to the 2019-08-01 API version). The 204 is then removed (or not added) in the already existing 2019-08-15 API version). The comments in the PR also indicates that this is a new API version that no-one is using. But that is only true w. regards to the 2019-06-04 API version not previously being documented. Presumably other API versions (both later and earlier) are in use by clients.

At this point, I'm not sure what is intentional and what changes are accidental.

tushar-bhargava-ms commented:
The primary role of this PR is to introduce the `tagsList` field in our Swagger spec from API versions 2019-06-04 and onwards. The addition of this field to our Swagger spec as well as the addition of the 2019-06-04 API version had been overlooked accidentally, and this PR was meant to fix that omission.

However, when I submitted this PR, I realized that new automated checks had been introduced in this repository and that these automated checks were failing because of syntax errors in the Gallery Application part of the Swagger spec. (I had simply duplicated an existing API version Swagger spec for creating the 2019-06-04 API version.)

Once I was aware of the syntax error in the Gallery Application part of the Swagger spec, I decided to remedy that error in our other specs, which led to the introduction of several new files in the PR.

The Gallery Application provider is administered by a completely different team, and they have no users at the moment as Allison said. I have been treating their part of the spec as more of a placeholder, and it is very likely that there'll be subsequent PRs from their side to improve their part of the spec. However, for the time being, I just wanted to fix the syntax error in their part of the spec so that all automated checks for this PR pass, and I can merge my `tagsList`-related changes.

johanste commented:
I see - it is very unfortunate if we have descriptions of APIs that a) are not supposed to be used by customers and b) are essentially work in progress in the master branch of the repository. The correct solution would be to remove the APIs that are either not ready for consumption or not yet implemented from the documents in the master branch. Is this an option?

tushar-bhargava-ms commented:
I think that might be a viable option, but I'd have to discuss it with my tech lead first. Let me get back to you on this.

tushar-bhargava-ms committed:
Removed Gallery Application schema from 2019-08-01 directory.

tushar-bhargava-ms committed:
Removed Gallery Application provider from 2019-06-01 spec.

tushar-bhargava-ms committed:
Removed Gallery Application provider from 2019-06-04 spec.

tushar-bhargava-ms committed:
Removing Gallery Application provider spec from 2019-08-15 directory.

tushar-bhargava-ms commented:
@johanste  Good news! I discussed the issue with my tech lead and got the go ahead. I've removed the troublesome Gallery Application provider schema from the spec. I've also updated the PR title and description to be clearer.

tushar-bhargava-ms commented:
@johanste @Juliehzl Sorry for the repeated pings, but if you could review and approve this PR by this week, I'd really appreciate it! Our customers are waiting on the newest Swagger specs.

johanste commented:
No API change - this is correcting an existing swagger. I see no issues with the addition.

Juliehzl commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2019-06-04/imds.json:
```diff
@@ -0,0 +1,828 @@
+                        "in": "query",
+                        "type": "string",
+                        "description": "This is the urlencoded identifier URI of the sink resource for the requested Azure AD token. The resulting token contains the corresponding aud for this resource.",
+                        "required": true
```
Why this parameter is required in query? What will happen when users don't provide `resource` parameter?

tushar-bhargava-ms replied:
The token returned by this query is specific to a resource. Without this parameter, the query will fail.

Juliehzl commented:
LGTM.
